tests/scenario/test_machine_charm/test_cos_agent_e2e.py

```diff
@@ -94,50 +103,119 @@ def __init__(self, framework: Framework):
                 refresh_events=[self.on.cos_agent_relation_changed],
             )
 
+    # Step 1: gagent is deployed and ends in "unknown" state
+    # since it is not related to the principal charm.
     cos_agent = Relation("cos-agent")
-
-    state_out = State(
-        relations=[
-            cos_agent,
-        ]
-    ).trigger(
-        cos_agent.changed_event(remote_unit_id=1),
+    ctx = Context(
         charm_type=MyPrincipal,
         meta=MyPrincipal.META,
         charm_root=vroot,
     )
 
-    # step 2: gagent is notified that the principal has touched its relation data
+    state = State(relations=[cos_agent])
+    state_out = ctx.run(cos_agent.changed_event(remote_unit_id=1), state=state)
+    assert state_out.status.unit.name == "unknown"
+
+    # Step 2: gagent is related to principal charm and ends in "blocked" status
+    # since there are missing relations:
+    #  - send-remote-write
+    #  - logging-consumer
+    #  - grafana-dashboards-provider
     peer = PeerRelation("peers")
     cos_agent1 = SubordinateRelation(
         "cos-agent",
         primary_app_name="mock-principal",
         remote_unit_data=state_out.relations[0].local_unit_data,
     )
-    state_out1 = State(relations=[cos_agent1, peer]).trigger(
-        cos_agent1.changed_event(remote_unit_id=0),
+
+    ctx1 = Context(
```

Comment:
you shouldn't need to reinstantiate the context every time. In fact I typically add a single `ctx` fixture at the top of the test file (or even in a `conftest.py`) and reuse it throughout.